subprojects/core/src/main/java/org/gradle/api/internal/project/taskfactory/TaskIdentity.java

```diff
@@ -51,14 +51,37 @@
         this.uniqueId = uniqueId;
     }
 
+    /**
+     * Create a task identity.
+     */
     public static <T extends Task> TaskIdentity<T> create(String name, Class<T> type, ProjectInternal project) {
+        return doCreate(
+            name,
+            type,
+            project,
+            SEQUENCE.getAndIncrement()
+        );
+    }
+
+    /**
+     * Create a task identity.
+     *
+     * Should only be used when loading from the configuration cache to preserve task ids.
+     */
+    public static <T extends Task> TaskIdentity<T> create(String name, Class<T> type, ProjectInternal project, long uniqueId) {
+        // Increment the sequence so later assigned task ids don't clash with the one manually assigned.
+        SEQUENCE.getAndAccumulate(uniqueId, (current, givenId) -> current > givenId ? current : givenId + 1);
```

Comment:
❓ Could there be a case where we would have created tasks via other methods and only then load others from the cache? If yes, are there ways to protect from this?

Potential bad case:
- A new task is created via the other `create` and assigned id of `0`.
- This `create` method gets called with the `0` id loaded from cache
- We end up having two tasks with the same id `0` that should have been unique.

Even if it is currently impossible, could we make an assumption that all the `create(..., uniqueId)` calls happen before any other `create(...)` calls and add at least a warning here or maybe actually fail? The sanity check for this would be something like `SEQUENCE.get() <= uniqueId`.

Reply:
This will also require (and guarantee) that when tasks are loaded from the cache they are loaded in the deterministic order (ascending `uniqueId`)

Reply:
Currently this isn't possible, since using the call with unique id always happens after all the other calls. We should also codify this assumption, though let's do it in a follow up PR. The basic idea there would be that you'd use a service for creating unique ids that can't be called after you selected the first id manually.